src/source/tcp.rs

```diff
@@ -151,23 +194,28 @@ where
         loop {
             match listener.accept() {
                 Ok((stream, _addr)) => {
+                    // Actually spawn the stream handler
                     let rchans = chans.to_owned();
-                    let new_stream = thread::spawn(move |poller| {
-                        // Note - Stream handlers are allowed to crash without
-                        // compromising Cernan's ability to gracefully shutdown.
-                        poller
-                            .register(
-                                &stream,
-                                mio::Token(0),
-                                mio::Ready::readable(),
-                                mio::PollOpt::edge(),
-                            )
-                            .unwrap();
-
-                        let mut handler = H::new();
-                        handler.handle_stream(rchans, &poller, stream);
-                    });
-                    self.handlers.push(new_stream);
+                    let stream_events_readiness = self.stream_events_readiness.clone();
+                    let new_stream = thread::spawn2(
+                        stream_events_readiness,
+                        move |poller| {
+                            // Note - Stream handlers are allowed to crash without
+                            // compromising Cernan's ability to gracefully shutdown.
+                            poller
+                                .register(
+                                    &stream,
+                                    mio::Token(0),
+                                    mio::Ready::readable(),
+                                    mio::PollOpt::edge(),
+                                )
+                                .unwrap();
+
+                            let mut handler = H::new();
+                            handler.run(rchans, &poller, stream);
+                        });
+                    let handler_id = self.handlers.len();
```

Comment:
Subtle bug here; this needs to be a monotonically increasing counter. 

For example:
Imagine you have 4 clients, then the first 2 disconnect. The next 2 connections will get the same value as the remaining 2.